### #1. Importing libraries

In [4]:
import pandas as pd
import numpy as np
import os

In [5]:
pathname=r'/Users/alejandrocisneros/Desktop/Data Analysis Coursework/Data Immersion/Achievement 4/02-2024 Instacart Basket Analysis'

#### #Testing pathname functionality

In [6]:
pathname

'/Users/alejandrocisneros/Desktop/Data Analysis Coursework/Data Immersion/Achievement 4/02-2024 Instacart Basket Analysis'

In [7]:
df_orders=pd.read_csv(os.path.join(pathname, '2. Data', 'Original Data', 'orders.csv'),index_col=False)

In [8]:
df_products=pd.read_csv(os.path.join(pathname, '2. Data', 'Original Data', 'products.csv'),index_col=False)

##### #Both dataframes (df) have been established using OS path functionality

### #2. Data Wrangling Initiation

In [9]:
#dropping the eval set column from orders.csv
df_orders=df_orders.drop(columns=['eval_set'])

#### #If we want to avoid re-writing the dataset with the command above, you can always create a copy of the dataset in which you remove the eval_set column like this:
df_orders_2 = df_orders.drop(columns = ['eval_set'])

In [10]:
df_orders['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

#### #Now, we will rename the dow column to something more intuitive

In [11]:
df_orders.rename(columns={'order_dow': 'orders_day_of_week'}, inplace=True)

#### #Testing the functionality of code above

In [12]:
df_orders.head(5)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### #Sucess!
#### #Now, in order to calculate descriptive statistics and exclude columns to which these statistics dont apply, we will change the data type of the user_id and order_id columns to string (or O in pandas).

In [13]:
df_orders['order_id']=df_orders['order_id'].astype('str')

In [14]:
df_orders['user_id']=df_orders['user_id'].astype('str')

In [15]:
df_orders['order_id'].dtype

dtype('O')

In [16]:
df_orders['user_id'].dtype

dtype('O')

In [17]:
df_dep=pd.read_csv(os.path.join(pathname, '2. Data', 'Original Data', 'departments.csv'),index_col=False)

In [18]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

#### #Here, we will transpose this table and transition from wide to long format.

In [19]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### #Now, instead of re-writing the original data frame with the new format, we are creating a new data frame in which the departments table is long format as opposed to the original wide format.

In [20]:
df_dep_t=df_dep.T

In [21]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### #New dataframe, df_dep_t, was successfully created.

#### #Now, we will reset the strange looking index and have it say 'department_id' on the left column header and 'department' on the right column header.

In [22]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

#### #Now that you have an index column, let’s create a new header for your dataframe. This will be done in three steps: 1) Create a new header; 2) Remove the first row in the dataframe; and 3) Add the new header. This all sounds a little abstract, so let’s walk through the process together.

In [23]:
# Take the first row of the df_dep_t as the new header
# iloc is a function used to assign values designated by the index number of a row or column
new_header=df_dep_t.iloc[0]

In [24]:
new_header

0    department
Name: department_id, dtype: object

In [25]:
# Take the data under the header row for a new df
df_dep_t_new=df_dep_t[1:]

In [26]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [29]:
# This allows for the recently created new_header variable to be used as the new column headers
df_dep_t_new.columns=new_header

In [30]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#### #Now, we will turn this departments table into our data dictionary

In [31]:
data_dictionary=df_dep_t_new.to_dict('index')

In [32]:
data_dictionary

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [33]:
df_products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

#We can now print the specific dictionary key associated with each department id and have an idea of what department each product belongs to.

In [35]:
print(data_dictionary.get('19'))

{'department': 'snacks'}


In [37]:
# Now, we create a subset in which we are only looking at products within the snacks category
df_snacks = df_products[df_products['department_id']==19]

In [39]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [40]:
#This is another way of creating a subset but using the 'loc' function instead
df_snacks_2=df_products.loc[df_products['department_id']==19]

In [43]:
#If you want to look in more than one column, using 'loc' function with 'isin' function allows for that to happen.
#Instead of using the equal sign to find one specific column, 'isin' allows you to look in a list and pick columns.
df_snacks_3=df_products.loc[df_products['department_id'].isin([17, 18, 19])]

In [45]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [46]:
df_snacks_3

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
13             14                 Fresh Scent Dishwasher Cleaner        74   
14             15                       Overnight Diapers Size 6        56   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
...           ...                                            ...       ...   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49672       49668                 Apple Cinnamon Scented Candles       101   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
13                17     6.5  
14                18    11.2  
15                19     5.2  
24                19     1.9  
...              ...     ...  
49669             19     6.9  
49670             19     5.2  
49672             17     5.6  
49675             19     3.0  
49683             19     6.0  

[10430 rows x 5 columns]

In [47]:
df_orders.to_csv(os.path.join(pathname, '2. Data', 'Original Data', 'orders_wrangled.csv'))

## #3. Ending of in-course work

## #4.Exercise 4.4 procedures

In [51]:
df_orders.head(50)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   
11  2168274       2             1                   2                 11   
12  1501582       2             2                   5                 10   
13  1901567       2             3                   1                 10   
14   738281       2             4                   2                 10   
15  1673511       2             5                   3                 11   
16  1199898       2             6                   2                  9   
17  3194192       2             7                   2                 12   
18   788338       2             8                   1                 15   
19  1718559       2             9                   2                  9   
20  1447487       2            10                   1                 11   
21  1402090       2            11                   1                 10   
22  3186735       2            12                   1                  9   
23  3268552       2            13                   4                 11   
24   839880       2            14                   3                 10   
25  1492625       2            15                   1                 11   
26  1374495       3             1                   1                 14   
27   444309       3             2                   3                 19   
28  3002854       3             3                   3                 16   
29  2037211       3             4                   2                 18   
30  2710558       3             5                   0                 17   
31  1972919       3             6                   0                 16   
32  1839752       3             7                   0                 15   
33  3225766       3             8                   0                 17   
34  3160850       3             9                   0                 16   
35   676467       3            10                   3                 16   
36   521107       3            11                   0                 18   
37  1402502       3            12                   1                 15   
38  2774568       3            13                   5                 15   
39  3343014       4             1                   6                 11   
40  2030307       4             2                   4                 11   
41   691089       4             3                   4                 15   
42    94891       4             4                   5                 13   
43  2557754       4             5                   5                 13   
44   329954       4             6                   3                 12   
45  2717275       5             1                   3                 12   
46  1909121       5             2                   0                 16   
47  2267326       5             3                   3                 18   
48   157374       5             4                   1                 18   
49  2196797       5             5                   0                 11   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3         

### #2. Question 2

#### #Changing order_number data type from integer/float to string so as to exclude it in the descriptive analyses

In [52]:
df_orders['order_number']=df_orders['order_number'].astype('str')

In [53]:
df_orders['order_number'].dtype

dtype('O')

In [57]:
#Testing descriptive analysis to see if data type change succeeded
df_orders.describe()

orders_day_of_week  order_hour_of_day  days_since_last_order
count        3.421083e+06       3.421083e+06           3.214874e+06
mean         2.776219e+00       1.345202e+01           1.111484e+01
std          2.046829e+00       4.226088e+00           9.206737e+00
min          0.000000e+00       0.000000e+00           0.000000e+00
25%          1.000000e+00       1.000000e+01           4.000000e+00
50%          3.000000e+00       1.300000e+01           7.000000e+00
75%          5.000000e+00       1.600000e+01           1.500000e+01
max          6.000000e+00       2.300000e+01           3.000000e+01

#### #Data type reformatting SUCCESS

### #Question 3

#### #Renaming the days_since_prior_order column to days_since_last_order to make it a little more intuitive

In [54]:
df_orders.rename(columns={'days_since_prior_order': 'days_since_last_order'}, inplace=True)

In [56]:
# Checking for results
df_orders.head(5)

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

#### #4. Question 4
### #Here, we will create a variable that finds the most common hour of order placement.

In [58]:
popular_hour=df_orders['order_hour_of_day'].mode()

In [59]:
popular_hour

0    10
Name: order_hour_of_day, dtype: int64

#### This shows that the hour in which most orders were placed was 10 am

### #5. Question 5

In [60]:
#Here, we are determining what classification name for "4" in the category_id column by accessing our data dictionary
print(data_dictionary.get('4'))

{'department': 'produce'}


#### This shows us the answer. Where 4=category_id, the product category is produce

### #6. Question 6

In [62]:
#Here, we are creating a subset that contains information only on products from the 'breakfast' category
df_breakfast = df_products[df_products['department_id']==14]

In [63]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

#### #Below, we are checking our current dataframes and subsets.

In [64]:
df_orders.head()

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [65]:
df_products.head()


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [66]:
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

In [67]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### #7. Question 7

#### Above, we checked the dataframes we have altered/created to keep track of them. Below, we will create another subset in which we focus on items from the alcohol, deli, beverages and meat/seafood departments.

In [70]:
df_dinner_party=df_products[df_products['department_id'].isin([5, 7, 12, 20])]

In [76]:
#Checking for results
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### #8. Question 8
#### Last data frame we created had 7,650 rows

### #9. Question 9

#### #9. We need to create a new dataframe for where user_id=1

In [80]:
#Creating a data frame (Quotations around 1 given that we changed the data type for the user_id column to 'str')
df_orders_user_1=df_orders[df_orders['user_id']=='1']

In [81]:
df_orders_user_1

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

### #10. Question 10

#### #10. The following stats give us all the information we can get on order details from user_id = 1

In [88]:
df_orders_user_1.dtypes

order_id                  object
user_id                   object
order_number              object
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

In [82]:
df_orders_user_1.shape

(11, 6)

In [83]:
df_orders_user_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11 entries, 0 to 10
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               11 non-null     object 
 1   user_id                11 non-null     object 
 2   order_number           11 non-null     object 
 3   orders_day_of_week     11 non-null     int64  
 4   order_hour_of_day      11 non-null     int64  
 5   days_since_last_order  10 non-null     float64
dtypes: float64(1), int64(2), object(3)
memory usage: 616.0+ bytes


In [84]:
df_orders_user_1.describe

<bound method NDFrame.describe of    order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_last_order  
0                     NaN  
1        

In [85]:
df_orders_user_1.describe()

orders_day_of_week  order_hour_of_day  days_since_last_order
count           11.000000          11.000000              10.000000
mean             2.636364          10.090909              19.000000
std              1.286291           3.477198               9.030811
min              1.000000           7.000000               0.000000
25%              1.500000           7.500000              14.250000
50%              3.000000           8.000000              19.500000
75%              4.000000          13.000000              26.250000
max              4.000000          16.000000              30.000000

In [87]:
df_orders_user_1.describe(include='all')

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
count        11      11           11           11.000000          11.000000   
unique       11       1           11                 NaN                NaN   
top     2539329       1            1                 NaN                NaN   
freq          1      11            1                 NaN                NaN   
mean        NaN     NaN          NaN            2.636364          10.090909   
std         NaN     NaN          NaN            1.286291           3.477198   
min         NaN     NaN          NaN            1.000000           7.000000   
25%         NaN     NaN          NaN            1.500000           7.500000   
50%         NaN     NaN          NaN            3.000000           8.000000   
75%         NaN     NaN          NaN            4.000000          13.000000   
max         NaN     NaN          NaN            4.000000          16.000000   

        days_since_last_order  
count               10.000000  
unique                    NaN  
top                       NaN  
freq                      NaN  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [90]:
df_orders.to_csv(os.path.join(pathname, '2. Data', 'Prepared Data', 'orders_wrangled.csv'))

In [91]:
df_dep_t_new.to_csv(os.path.join(pathname, '2. Data', 'Prepared Data', 'departments_wrangled.csv'))